quartz/components/Head.tsx

```diff
@@ -1,21 +1,146 @@
 import { i18n } from "../i18n"
 import { FullSlug, joinSegments, pathToRoot } from "../util/path"
 import { JSResourceToScriptElement } from "../util/resources"
-import { QuartzComponent, QuartzComponentConstructor, QuartzComponentProps } from "./types"
+import { QuartzComponentConstructor, QuartzComponentProps } from "./types"
+import satori, { SatoriOptions } from "satori"
+import * as fs from "fs"
+import { ImageOptions, SocialImageOptions, getSatoriFont } from "../util/imageHelper"
+import sharp from "sharp"
+import { defaultImage } from "../util/socialImage"
+import { unescapeHTML } from "../util/escape"
+
+/**
+ * Generates social image (OG/twitter standard) and saves it as `.webp` inside the public folder
+ * @param opts options for generating image
+ */
+async function generateSocialImage(
+  { cfg, description, fileName, fontsPromise, title }: ImageOptions,
+  userOpts: SocialImageOptions,
+) {
+  const fonts = await fontsPromise
+
+  // JSX that will be used to generate satori svg
+  const imageElement = userOpts.imageStructure(cfg, userOpts, title, description, fonts)
+
+  const svg = await satori(imageElement, {
+    width: userOpts.width,
+    height: userOpts.height,
+    fonts: fonts,
+  })
+
+  // Convert svg directly to webp (with additional compression)
+  const compressed = await sharp(Buffer.from(svg)).webp({ quality: 40 }).toBuffer()
+
+  // Write to file system
+  fs.writeFileSync(`${imageDir}/${fileName}.${extension}`, compressed)
+}
+
+const extension = "webp"
+const imageDir = "public/static/social-images"
+
+const defaultOptions: SocialImageOptions = {
+  colorScheme: "lightMode",
+  width: 1200,
+  height: 676,
+  imageStructure: defaultImage,
+}
 
 export default (() => {
-  const Head: QuartzComponent = ({ cfg, fileData, externalResources }: QuartzComponentProps) => {
-    const title = fileData.frontmatter?.title ?? i18n(cfg.locale).propertyDefaults.title
-    const description =
+  let fontsPromise: Promise<SatoriOptions["fonts"]>
+
+  let fullOptions: SocialImageOptions
+  function Head({ cfg, fileData, externalResources }: QuartzComponentProps) {
+    // Initialize options if not set
+    if (!fullOptions) {
+      if (typeof cfg.generateSocialImages !== "boolean") {
+        fullOptions = { ...defaultOptions, ...cfg.generateSocialImages }
+      } else {
+        fullOptions = defaultOptions
+      }
+    }
+
+    // Memoize google fonts
+    if (!fontsPromise) {
+      fontsPromise = getSatoriFont(cfg.theme.typography.header, cfg.theme.typography.body)
+    }
+
+    const slug = fileData.filePath
```

Comment:
ah you should just use fileData.slug here i think.
```suggestion
    const slug = fileData.slug!
```

Reply:
I think you brought this up earlier here: https://github.com/jackyzha0/quartz/pull/740#issuecomment-1929249117
Should be `filePath` to ensure we have a unique file path, I don't think slugs have to be globally unique